[EventDispatcher][VarDumper] optimize perf by leveraging Closure::fromCallable()

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Callables are notably slower than closures. Let's turn then to closures thanks to `Closure::fromCallable()`.

This doesn't affect performance for run-once listeners.
And improves performance for events that are dispatched several times.

Same for VarDumper's casters.

nicolas-grekas committed:
[EventDispatcher][VarDumper] optimize perf by leveraging Closure::fromCallable()

ostrolucky commented on src/Symfony/Component/EventDispatcher/EventDispatcher.php:
```diff
@@ -233,4 +247,28 @@ private function sortListeners($eventName)
+        foreach ($this->listeners[$eventName] as &$listeners) {
+            foreach ($listeners as &$listener) {
+                $closure = &$optimized[];
+                $closure = $listener instanceof \Closure ? $listener : static function (...$args) use (&$listener, &$closure) {
```
I don't think this intermediary closure is needed

nicolas-grekas replied:
this closure allows creating the listener services lazily: right now, all listener services are created even if one of them stops propagation

ostrolucky replied:
Yes but listeners rarely stop propagation. I think we should avoid doing these "cache misses" and assume most listeners are going to be executed.

nicolas-grekas replied:
why? to save the closure? it doesn't have any impact. Laziness is much more desirable to me here.

nicolas-grekas replied:
oh, actually we can move the "if" outside of the closure and save creating a closure when no lazy listeners are involved. updated.

ostrolucky reviewed (changes requested):
Why is this not incorporated inside sortListeners function? Instead it repeats the logic but ignores sort marking and is used only during dispatching

> This doesn't affect performance for run-once listeners.

How come? It makes extra operations (new closure, Closure::fromCallable call)

nicolas-grekas commented on src/Symfony/Component/EventDispatcher/EventDispatcher.php:
```diff
@@ -30,6 +30,14 @@ class EventDispatcher implements EventDispatcherInterface
+
+    public function __construct()
+    {
+        if (__CLASS__ === \get_class($this)) {
```
because this optimization is internal: the `getListeners()` should continue returning the exact listeners that where added - doing otherwise would be a BC break

ostrolucky replied:
getListeners() is currently completely detached from optimizeListeners()

nicolas-grekas replied:
`dispatch()` calls `getListeners()`: in case the class is overridden, it should continue calling it.
This means we cannot call `optimizeListeners()` for child classes.

ostrolucky replied:
Well that sucks. Can we deprecate calling `getListeners` from `dispatch` by passing special flag?

nicolas-grekas replied:
Not in this PR at least :)

nicolas-grekas commented on src/Symfony/Component/EventDispatcher/EventDispatcher.php:
```diff
@@ -40,7 +48,13 @@ public function dispatch($eventName, Event $event = null)
+            $listeners = $this->getListeners($eventName);
+        }
+
+        if ($listeners) {
```
saves a method call and was already there

nicolas-grekas commented on src/Symfony/Component/EventDispatcher/EventDispatcher.php:
```diff
@@ -123,7 +137,9 @@ public function hasListeners($eventName = null)
     {
         $this->listeners[$eventName][$priority][] = $listener;
-        unset($this->sorted[$eventName]);
+        if (isset($this->sorted[$eventName]) || isset($this->optimized[$eventName])) {
```
for perf: "unset" is 3-4 times slower that doing this "if", even on empty arrays

ostrolucky replied:
Not according my test

```
>>> timeit -n10000 '$b[0] = 1;unset($b[0];'
=> "$b[0] = 1;unset($b[0];"
Command took 0.000001 seconds on average (0.000001 median; 0.012583 total) to complete.
>>> timeit -n10000 '$b[0] = 1;if (isset($b[0])) unset($b[0]);'
=> "$b[0] = 1;if (isset($b[0])) unset($b[0]);"
Command took 0.000001 seconds on average (0.000001 median; 0.012581 total) to complete.
>>> timeit -n10000 '$b = [];unset($b[0];'
=> "$b = [];unset($b[0];"
Command took 0.000001 seconds on average (0.000001 median; 0.012470 total) to complete.
>>> timeit -n10000 '$b = [];if (isset($b[0])) unset($b[0]);'
=> "$b = [];if (isset($b[0])) unset($b[0]);"
Command took 0.000001 seconds on average (0.000001 median; 0.012692 total) to complete.
>>> 
```

nicolas-grekas replied:
Here is mine: 12ms vs 4ms

<details>

```php
<?php
  
namespace bench;

$iter = 100000;

function a($i)
{
    $a = $b = array();
    while (--$i) {
        unset($a['abc'], $b['abc']);
    }
}

function b($i)
{
    $a = $b = array();
    while (--$i) {
        if (isset($a['abc']) || isset($b['abc'])) {
            unset($a['abc'], $b['abc']);
        }
    }
}

$s = microtime(1);
a($iter);

echo 'a',1000*(microtime(true)-$s), "\n";

$s = microtime(1);
b($iter);

echo 'b',1000*(microtime(true)-$s), "\n";
```

</details>

ostrolucky replied:
According to that benchmark `a()` is consistently faster https://3v4l.org/NlGYf

nicolas-grekas replied:
Wow, thanks for double checking. I had xdebug enabled, that's why!

nicolas-grekas commented on src/Symfony/Component/EventDispatcher/EventDispatcher.php:
```diff
@@ -233,4 +247,28 @@ private function sortListeners($eventName)
+            foreach ($listeners as &$listener) {
+                $closure = &$optimized[];
+                $closure = $listener instanceof \Closure ? $listener : static function (...$args) use (&$listener, &$closure) {
+                    if (\is_array($listener) && $listener[0] instanceof \Closure) {
```
isset added

nicolas-grekas commented:
@ostrolucky thanks for the review!

> Why is this not incorporated inside sortListeners function? Instead it repeats the logic but ignores sort marking and is used only during dispatching

Because sortListeners has public side effect: it drive the return value of `getListeners()`.
Changing would mean breaking BC.

> How come? It makes extra operations (new closure, Closure::fromCallable call)

and removes some. I did benchmark this, there is no measurable difference.

ostrolucky commented:
> it drive the return value of getListeners()

> getListeners() should continue returning the exact listeners that where added

I didn't mean to to imply you should save the optimization result into `sorted` property, you can still put it to `optimized`, but from inside sortListeners. That wouldn't change anything on public side.

nicolas-grekas commented:
> you can still put it to optimized, but from inside sortListeners

that could defeat the extra laziness added here (not instantiating lazy listeners when propagation is stopped before them)

ostrolucky commented:
> that could defeat the extra laziness added here (not instantiating lazy listeners when propagation is stopped before them)

I don't see how would it defeat it :thinking: Only way that closure is replaced with delegated one is when it's executed, which will happen only during dispatching.

nicolas-grekas commented:
Because of the `$listener[0] = $listener[0]();` line, which is the one that is made lazy here.

ostrolucky commented:
I get that, but that line is executed only when outer closure is executed. I don't see why would moving construction of this closure to different place change when is it executed

nicolas-grekas commented:
Because we have to call that line to populate the sortedListeners property.

ostrolucky commented on src/Symfony/Component/EventDispatcher/EventDispatcher.php:
```diff
@@ -215,22 +224,46 @@ protected function doDispatch($listeners, $eventName, Event $event)
+                $closure = &$this->optimized[$eventName][];
+                if (\is_array($listener) && isset($listener[0]) && $listener[0] instanceof \Closure) {
+                    $closure = static function (...$args) use (&$listener, &$closure) {
+                        if ($listener[0] instanceof \Closure) {
```
Seems this if is unnecessary now

nicolas-grekas replied:
It is: the value can change by reference before the closure is called.

ostrolucky replied:
in that case is_array and isset are also needed?

nicolas-grekas replied:
not really, because the only mutations that can happen are the one done in the same class, which only resolve the listener - never more.

ostrolucky commented:
This closure trick increases speed by 0.6μs (8%). Measured with phpbench, 100 listeners attached (and all of them executed). Correlation on number of listeners is very linear.

I don't think this is worth increased complexity and number of untested paths. Also, is there a PHP bug report? I think this is something better to be improved on language level.

nicolas-grekas commented:
We made current Symfony versions really fast because we made such changes. Small improvements for your bench can be a significant one for a specific use case. And it's already been provided that many small improvements aggregate to significant general ones.

The real issue is having EventDispatcher. We should inspect existing use cases where this is needed (if any) and consider deprecating if possible IMHO. But that shouldn't block this one.

ostrolucky commented:
Well if we decide it's worth it, can you at least properly comment this code, cover it by tests and notify upstream about workarounds we need to do to increase performance so there is at least some chance we can eventually remove them? This was proven successful with eg inlining constants in DI Container. Performance tweaks like this are killing readability and hence ability to debug and contribute a lot. As you can see from my questions there are lot of non-obvious things for other people.

I still don't know in what case would `if ($listener[0] instanceof \Closure) {` condition be needed, because it's not covered by tests. And there is no longer even coverage for calling getListeners() call from dispatch().

stof commented:
> I still don't know in what case would `if ($listener[0] instanceof \Closure) {` condition be needed, because it's not covered by tests.

that part is the lazy-loading of the listener object.

nicolas-grekas commented:
The discussed code path is now tested @ostrolucky, testMutatingWhilePropagationIsStopped.

Don't be fooled by the diff stat, I just renamed `AbstractEventDispatcherTest` to `EventDispatcherTest` and added this test case.

ostrolucky commented:
Thanks. I can see now that same closure is evaluated by sortListeners and optimizeListeners might not re-triggered after that happens.
